extension/observer/hostobserver/extension.go

```diff
@@ -0,0 +1,164 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hostobserver
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"syscall"
+
+	"github.com/shirou/gopsutil/net"
+	"github.com/shirou/gopsutil/process"
+	"go.opentelemetry.io/collector/component"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer"
+)
+
+type hostObserver struct {
+	observer.EndpointsWatcher
+	logger *zap.Logger
+}
+
+var _ component.ServiceExtension = (*hostObserver)(nil)
+
+func newObserver(logger *zap.Logger, config *Config) (component.ServiceExtension, error) {
+	h := &hostObserver{
+		logger: logger,
+		EndpointsWatcher: observer.EndpointsWatcher{
+			RefreshInterval: config.RefreshInterval,
+			ListEndpoints:   (&hostObserver{}).listEndpoints,
+		},
+	}
+
+	return h, nil
+}
+
+func (h *hostObserver) Start(context.Context, component.Host) error {
+	return nil
+}
+
+func (h *hostObserver) Shutdown(context.Context) error {
+	h.StopListAndWatch()
+	return nil
+}
+
+func (h *hostObserver) listEndpoints() []observer.Endpoint {
+
+	var conns []net.ConnectionStat
+	var err error
+	// Skip UID lookup since it's not used by the observer, the method
+	// is available only on linux. See https://github.com/shirou/gopsutil/pull/783
+	// for details.
+	if runtime.GOOS == "linux" {
+		conns, err = net.ConnectionsWithoutUids("all")
+	} else {
+		conns, err = net.Connections("all")
+	}
+
+	if err != nil {
+		h.logger.Error("Could not get local network listeners", zap.Error(err))
+	}
+
+	endpoints := make([]observer.Endpoint, 0, len(conns))
+	connsByPID := make(map[int32][]*net.ConnectionStat)
+	for i := range conns {
+		c := conns[i]
+		isIPSocket := c.Family == syscall.AF_INET || c.Family == syscall.AF_INET6
+		isTCPOrUDP := c.Type == syscall.SOCK_STREAM || c.Type == syscall.SOCK_DGRAM
+		// UDP doesn't have any status
+		isUDPOrListening := c.Type == syscall.SOCK_DGRAM || c.Status == "LISTEN"
+		// UDP is "listening" when it has a remote port of 0
+		isTCPOrHasNoRemotePort := c.Type == syscall.SOCK_STREAM || c.Raddr.Port == 0
+
+		// PID of 0 means that the listening file descriptor couldn't be mapped
+		// back to a process's set of open file descriptors in /proc
+		if !isIPSocket || !isTCPOrUDP || !isUDPOrListening || !isTCPOrHasNoRemotePort || c.Pid == 0 {
+			continue
```

Comment:
It might be useful behavior to still discover endpoints without the extra Caps and just omit the process information from the output.  Some people might just want to know if a port is listening and not care about process metadata.

Reply:
I'm not sure I follow. Could you please clarify?

Reply:
What happens if you don't have the right caps on Linux is that c.Pid will be 0, and thus you will completely ignore this listening socket.  But some people might want to know that the socket is listening and do discovery rules on them even though they don't get any of the process information.